apps/prairielearn/src/tests/accessibility/index.test.ts

```diff
@@ -19,47 +21,129 @@ const SITE_URL = 'http://localhost:' + config.serverPort;
+async function loadPageJsdom(url: string): Promise<{ text: string; jsdom: JSDOM }> {
   const text = await fetch(url).then((res) => {
     if (!res.ok) {
       throw new Error(`Error loading page: ${res.status}`);
     }
     return res.text();
   });
+
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+
+  let output = '';
+  const rewriter = new HTMLRewriter((chunk) => {
+    output += decoder.decode(chunk);
+  });
+
+  // We need to mimic what Bootstrap will do at runtime for tooltips and popovers:
+  // it will copy the `data-bs-title` attribute to the `aria-label` attribute if
+  // the element doesn't already have an `aria-label` attribute and if its text
+  // content is empty.
+  //
+  // Without this, we'll get a bunch of false positives that don't reflect
+  // the actual behavior at runtime.
+  rewriter.on('a, button', {
+    element(el) {
+      // This is slightly different than what we do at runtime. In practice, we'll
+      // only add an `aria-label` if the element is empty. But here, we can't
+      // inspect the element's children, so we'll only use the presence of an
+      // `aria-label` attribute to determine if we should add one.
+      if (el.hasAttribute('aria-label')) return;
+      const title = el.getAttribute('data-bs-title');
+      if (title) {
+        el.setAttribute('aria-label', title);
+      }
+    },
+  });
+
+  await rewriter.write(encoder.encode(text));
+  await rewriter.end();
+
   // JSDOM can be very verbose regarding unimplemented features (e.g., canvas).
   // We don't have a need to see these warnings, so we create a virtual console
   // that does not log anything.
   const virtualConsole = new VirtualConsole();
-  return new JSDOM(text, { virtualConsole });
+  return { text: output, jsdom: new JSDOM(output, { virtualConsole }) };
 }
 
 /**
  * Checks the given URL for accessibility violations.
  */
 async function checkPage(url: string) {
-  const page = await loadPageJsdom(SITE_URL + url);
-
-  // We need to mimic what Bootstrap will do at runtime for tooltips and popovers:
-  // it will copy the `data-bs-title` attribute to the `aria-label` attribute if
-  // the element doesn't already have an `aria-label` attribute and if its text
-  // content is empty.
-  //
-  // Without this, we'll get a bunch of false positives that don't reflect
-  // the actual behavior at runtime.
-  page.window.document.querySelectorAll('[data-bs-title]').forEach((el) => {
-    if (el.hasAttribute('aria-label') || el.textContent?.trim() !== '') {
-      return;
-    }
-    const title = el.getAttribute('data-bs-title');
-    if (title) {
-      el.setAttribute('aria-label', title);
-    }
+  const { text, jsdom } = await loadPageJsdom(SITE_URL + url);
+
+  let messages = '';
+
+  // Since the accessibility checks are already programmatically loading
+  // pretty much every page in the application, we'll piggyback on them
+  // to also run HTML validation.
```

Comment:
Maybe it's better to rename this collection of tests (and the corresponding directory and such) to "markup" tests?